File: src/server/bookings/__tests__/schemas.test.ts
import { describe, expect, it } from 'vitest'
import {
  availabilitySchema,
  bookSlotSchema,
  createBookingPageSchema,
  dateOverridesSchema,
  handleSchema,
  manageBookingSchema,
  publicAvailabilityQuerySchema,
  rescheduleSchema,
  slugSchema,
  timeRangeSchema,
  updateBookingPageSchema,
} from '#/server/bookings/schemas'

describe('handleSchema / slugSchema', () => {
  it('accepts valid handles', () => {
    expect(handleSchema.safeParse('anders').success).toBe(true)
    expect(handleSchema.safeParse('a-b-c').success).toBe(true)
    expect(handleSchema.safeParse('a1b').success).toBe(true)
  })

  it('rejects too short, too long, uppercase, leading/trailing hyphen, and invalid chars', () => {
    expect(handleSchema.safeParse('ab').success).toBe(false)
    expect(handleSchema.safeParse('a'.repeat(31)).success).toBe(false)
    expect(handleSchema.safeParse('Anders').success).toBe(false)
    expect(handleSchema.safeParse('-abc').success).toBe(false)
    expect(handleSchema.safeParse('abc-').success).toBe(false)
    expect(handleSchema.safeParse('ab_c').success).toBe(false)
    expect(handleSchema.safeParse('ab c').success).toBe(false)
  })

  it('slugSchema follows the same rules', () => {
    expect(slugSchema.safeParse('intro-call').success).toBe(true)
    expect(slugSchema.safeParse('IntroCall').success).toBe(false)
  })
})

describe('timeRangeSchema', () => {
  it('accepts 15-minute-aligned HH:mm with start < end', () => {
    expect(timeRangeSchema.safeParse({ start: '09:00', end: '09:15' }).success).toBe(true)
    expect(timeRangeSchema.safeParse({ start: '23:45', end: '23:45' }).success).toBe(false)
  })

  it('rejects unaligned minutes', () => {
    expect(timeRangeSchema.safeParse({ start: '09:05', end: '09:30' }).success).toBe(false)
    expect(timeRangeSchema.safeParse({ start: '09:00', end: '09:31' }).success).toBe(false)
  })

  it('rejects end <= start', () => {
    expect(timeRangeSchema.safeParse({ start: '10:00', end: '09:00' }).success).toBe(false)
  })

  it('rejects malformed hour/minute', () => {
    expect(timeRangeSchema.safeParse({ start: '24:00', end: '25:00' }).success).toBe(false)
    expect(timeRangeSchema.safeParse({ start: '9:00', end: '10:00' }).success).toBe(false)
  })
})

describe('availabilitySchema', () => {
  it('accepts weekday keys 0..6 with sorted, non-overlapping ranges', () => {
    const result = availabilitySchema.safeParse({
      '0': [],
      '1': [
        { start: '09:00', end: '12:00' },
        { start: '13:00', end: '17:00' },
      ],
    })
    expect(result.success).toBe(true)
  })

  it('rejects keys outside 0..6', () => {
    expect(availabilitySchema.safeParse({ '7': [] }).success).toBe(false)
    expect(availabilitySchema.safeParse({ monday: [] }).success).toBe(false)
    expect(availabilitySchema.safeParse({ '-1': [] }).success).toBe(false)
  })

  it('rejects overlapping ranges for a weekday', () => {
    const result = availabilitySchema.safeParse({
      '1': [
        { start: '09:00', end: '12:00' },
        { start: '11:00', end: '13:00' },
      ],
    })
    expect(result.success).toBe(false)
  })

  it('rejects unsorted ranges for a weekday', () => {
    const result = availabilitySchema.safeParse({
      '1': [
        { start: '13:00', end: '17:00' },
        { start: '09:00', end: '12:00' },
      ],
    })
    expect(result.success).toBe(false)
  })

  it('rejects more than 20 ranges in a day', () => {
    const ranges = Array.from({ length: 21 }, (_, i) => ({
      start: `${String(i).padStart(2, '0')}:00`,
      end: `${String(i).padStart(2, '0')}:15`,
    }))
    expect(availabilitySchema.safeParse({ '1': ranges }).success).toBe(false)
  })
})

describe('dateOverridesSchema', () => {
  it('accepts YYYY-MM-DD keys, empty array meaning day off', () => {
    const result = dateOverridesSchema.safeParse({
      '2026-12-24': [],
      '2026-12-31': [{ start: '09:00', end: '10:00' }],
    })
    expect(result.success).toBe(true)
  })

  it('rejects malformed date keys', () => {
    expect(dateOverridesSchema.safeParse({ '2026-1-1': [] }).success).toBe(false)
    expect(dateOverridesSchema.safeParse({ 'not-a-date': [] }).success).toBe(false)
  })

  it('rejects more than 366 entries', () => {
    const overrides: Record<string, []> = {}
    for (let i = 0; i < 367; i++) {
      const d = new Date(Date.UTC(2027, 0, 1) + i * 86_400_000)
      overrides[d.toISOString().slice(0, 10)] = []
    }
    expect(dateOverridesSchema.safeParse(overrides).success).toBe(false)
  })
})

const validPage = {
  slug: 'intro-call',
  title: '15 min intro',
  timezone: 'Europe/Oslo',
  slotDurationMin: 30,
  bufferBeforeMin: 0,
  bufferAfterMin: 0,
  minNoticeMin: 120,
  maxDaysAhead: 60,
  availability: { '1': [{ start: '09:00', end: '17:00' }] },
  googleSync: false,
  reminders: true,
}

describe('createBookingPageSchema', () => {
  it('accepts a valid page', () => {
    expect(createBookingPageSchema.safeParse(validPage).success).toBe(true)
  })

  it('rejects an invalid timezone', () => {
    const result = createBookingPageSchema.safeParse({ ...validPage, timezone: 'Not/AZone' })
    expect(result.success).toBe(false)
  })

  it('rejects out-of-range duration and buffers', () => {
    expect(createBookingPageSchema.safeParse({ ...validPage, slotDurationMin: 10 }).success).toBe(
      false,
    )
    expect(createBookingPageSchema.safeParse({ ...validPage, slotDurationMin: 481 }).success).toBe(
      false,
    )
    expect(createBookingPageSchema.safeParse({ ...validPage, bufferBeforeMin: 121 }).success).toBe(
      false,
    )
  })
})

describe('updateBookingPageSchema', () => {
  it('accepts a partial update with pageId and status', () => {
    const result = updateBookingPageSchema.safeParse({
      pageId: 'abc123',
      status: 'paused',
      title: 'New title',
    })
    expect(result.success).toBe(true)
  })

  it('requires pageId', () => {
    expect(updateBookingPageSchema.safeParse({ title: 'x' }).success).toBe(false)
  })
})

describe('publicAvailabilityQuerySchema', () => {
  const base = {
    handle: 'anders',
    slug: 'intro-call',
    timezone: 'Europe/Oslo',
  }

  it('accepts a window of 62 days or less', () => {
    expect(
      publicAvailabilityQuerySchema.safeParse({ ...base, from: '2026-01-01', to: '2026-03-03' })
        .success,
    ).toBe(true)
  })

  it('rejects a window greater than 62 days', () => {
    expect(
      publicAvailabilityQuerySchema.safeParse({ ...base, from: '2026-01-01', to: '2026-03-10' })
        .success,
    ).toBe(false)
  })

  it('rejects to before from', () => {
    expect(
      publicAvailabilityQuerySchema.safeParse({ ...base, from: '2026-03-10', to: '2026-01-01' })
        .success,
    ).toBe(false)
  })
})

describe('bookSlotSchema', () => {
  const base = {
    pageId: 'abc123',
    startAt: '2026-01-05T09:00:00.000Z',
    name: 'Ada Lovelace',
    timezone: 'Europe/Oslo',
  }

  it('requires email', () => {
    expect(bookSlotSchema.safeParse(base).success).toBe(false)
    expect(bookSlotSchema.safeParse({ ...base, email: 'not-an-email' }).success).toBe(false)
    expect(bookSlotSchema.safeParse({ ...base, email: 'ada@example.com' }).success).toBe(true)
  })
})

describe('manageBookingSchema / rescheduleSchema', () => {
  it('token is optional for the owner path', () => {
    expect(manageBookingSchema.safeParse({ bookingId: 'b1' }).success).toBe(true)
    expect(manageBookingSchema.safeParse({ bookingId: 'b1', token: 't' }).success).toBe(true)
  })

  it('rescheduleSchema requires a valid startAt', () => {
    expect(
      rescheduleSchema.safeParse({ bookingId: 'b1', startAt: '2026-01-05T09:00:00.000Z' }).success,
    ).toBe(true)
    expect(rescheduleSchema.safeParse({ bookingId: 'b1', startAt: 'not-a-date' }).success).toBe(
      false,
    )
  })
})

File: internal/bookings/schemas_test.go
package bookings_test

// Ports the PageInput-relevant cases from
// src/server/bookings/__tests__/schemas.test.ts's handleSchema/slugSchema, timeRangeSchema,
// availabilitySchema, dateOverridesSchema, and createBookingPageSchema describe blocks
// case-for-case. publicAvailabilityQuerySchema/bookSlotSchema/manageBookingSchema/
// rescheduleSchema/pageIdSchema belong to Task 3 (src/server/bookings/bookings.ts's own input
// types), not this task's PageInput.

import (
	"context"
	"errors"
	"fmt"
	"strings"
	"testing"
	"time"

	"github.com/refsdal/whenweall/internal/bookings"
)

func fieldsOf(t *testing.T, err error) map[string]string {
	t.Helper()
	var verr *bookings.ValidationError
	if !errors.As(err, &verr) {
		t.Fatalf("err = %v (%T), want *bookings.ValidationError", err, err)
	}
	return verr.Fields
}

func rangesOf(count int) []bookings.TimeRange {
	out := make([]bookings.TimeRange, count)
	for i := range out {
		out[i] = bookings.TimeRange{Start: fmt.Sprintf("%02d:00", i), End: fmt.Sprintf("%02d:15", i)}
	}
	return out
}

func TestPageInputHandleAndSlug(t *testing.T) {
	// PageInput.Slug shares handleSchema/slugSchema's HANDLE_SLUG_RE + 3..30 length rule; exercised
	// here through baseInput's Slug field and validateHandle (SetOrgSlug's own validator) side by
	// side, since both are the same underlying rule in the TS source.
	valid := []string{"anders", "a-b-c", "a1b", "intro-call"}
	for _, v := range valid {
		in := baseInput(func(in *bookings.PageInput) { in.Slug = v })
		if err := in.Validate(); err != nil {
			t.Errorf("Slug %q: Validate() = %v, want nil", v, err)
		}
	}

	invalid := []string{"ab", strings.Repeat("a", 31), "Anders", "-abc", "abc-", "ab_c", "ab c", "IntroCall"}
	for _, v := range invalid {
		in := baseInput(func(in *bookings.PageInput) { in.Slug = v })
		err := in.Validate()
		fields := fieldsOf(t, err)
		if _, ok := fields["slug"]; !ok {
			t.Errorf("Slug %q: Fields = %+v, want a slug entry", v, fields)
		}
	}
}

// TestValidateHandle exercises SetOrgSlug's own handle validation, which runs (and rejects a
// malformed handle) before any DB access — see SetOrgSlug's own doc comment in pages.go. Only the
// invalid cases are asserted here without a database; the valid path is covered behaviorally by
// TestSetOrgSlug in pages_test.go, which needs a real org row to update.
func TestValidateHandle(t *testing.T) {
	s := bookings.NewService(testConfig(t), nil)
	for _, v := range []string{"ab", strings.Repeat("a", 31), "Anders", "-abc", "abc-", "ab_c", "ab c"} {
		err := s.SetOrgSlug(context.Background(), "1", v)
		var verr *bookings.ValidationError
		if !errors.As(err, &verr) {
			t.Errorf("handle %q: err = %v, want *ValidationError", v, err)
		}
	}
}

func TestPageInputAvailability(t *testing.T) {
	t.Run("accepts weekday keys 0..6 with sorted, non-overlapping ranges, empty allowed", func(t *testing.T) {
		in := baseInput(func(in *bookings.PageInput) {
			in.Availability = bookings.Availability{
				"0": {},
				"1": {{Start: "09:00", End: "12:00"}, {Start: "13:00", End: "17:00"}},
			}
		})
		if err := in.Validate(); err != nil {
			t.Errorf("Validate() = %v, want nil", err)
		}
	})

	t.Run("rejects keys outside 0..6", func(t *testing.T) {
		for _, key := range []string{"7", "monday", "-1"} {
			in := baseInput(func(in *bookings.PageInput) {
				in.Availability = bookings.Availability{key: {}}
			})
			fields := fieldsOf(t, in.Validate())
			if _, ok := fields["availability"]; !ok {
				t.Errorf("key %q: Fields = %+v, want an availability entry", key, fields)
			}
		}
	})

	t.Run("rejects overlapping ranges for a weekday", func(t *testing.T) {
		in := baseInput(func(in *bookings.PageInput) {
			in.Availability = bookings.Availability{
				"1": {{Start: "09:00", End: "12:00"}, {Start: "11:00", End: "13:00"}},
			}
		})
		if err := in.Validate(); err == nil {
			t.Error("Validate() = nil, want an overlap error")
		}
	})

	t.Run("rejects unsorted ranges for a weekday", func(t *testing.T) {
		in := baseInput(func(in *bookings.PageInput) {
			in.Availability = bookings.Availability{
				"1": {{Start: "13:00", End: "17:00"}, {Start: "09:00", End: "12:00"}},
			}
		})
		if err := in.Validate(); err == nil {
			t.Error("Validate() = nil, want a sort error")
		}
	})

	t.Run("rejects more than 20 ranges in a day", func(t *testing.T) {
		in := baseInput(func(in *bookings.PageInput) {
			in.Availability = bookings.Availability{"1": rangesOf(21)}
		})
		if err := in.Validate(); err == nil {
			t.Error("Validate() = nil, want a too-many-ranges error")
		}
	})

	t.Run("rejects malformed or unaligned HH:mm, and end <= start", func(t *testing.T) {
		cases := [][2]string{
			{"09:05", "09:30"}, // unaligned minute
			{"09:00", "09:31"}, // unaligned minute
			{"10:00", "09:00"}, // end before start
			{"23:45", "23:45"}, // end == start
			{"24:00", "25:00"}, // malformed hour
			{"9:00", "10:00"},  // malformed (single-digit hour)
		}
		for _, c := range cases {
			in := baseInput(func(in *bookings.PageInput) {
				in.Availability = bookings.Availability{"1": {{Start: c[0], End: c[1]}}}
			})
			if err := in.Validate(); err == nil {
				t.Errorf("range %v: Validate() = nil, want an error", c)
			}
		}
	})
}

func TestPageInputDateOverrides(t *testing.T) {
	t.Run("accepts YYYY-MM-DD keys, empty slice meaning day off", func(t *testing.T) {
		in := baseInput(func(in *bookings.PageInput) {
			in.DateOverrides = bookings.DateOverrides{
				"2026-12-24": {},
				"2026-12-31": {{Start: "09:00", End: "10:00"}},
			}
		})
		if err := in.Validate(); err != nil {
			t.Errorf("Validate() = %v, want nil", err)
		}
	})

	t.Run("rejects malformed date keys", func(t *testing.T) {
		for _, key := range []string{"2026-1-1", "not-a-date"} {
			in := baseInput(func(in *bookings.PageInput) {
				in.DateOverrides = bookings.DateOverrides{key: {}}
			})
			fields := fieldsOf(t, in.Validate())
			if _, ok := fields["dateOverrides"]; !ok {
				t.Errorf("key %q: Fields = %+v, want a dateOverrides entry", key, fields)
			}
		}
	})

	t.Run("rejects more than 366 entries", func(t *testing.T) {
		overrides := make(bookings.DateOverrides, 367)
		start := time.Date(2027, 1, 1, 0, 0, 0, 0, time.UTC)
		for i := 0; i < 367; i++ {
			overrides[start.AddDate(0, 0, i).Format("2006-01-02")] = []bookings.TimeRange{}
		}
		in := baseInput(func(in *bookings.PageInput) { in.DateOverrides = overrides })
		fields := fieldsOf(t, in.Validate())
		if _, ok := fields["dateOverrides"]; !ok {
			t.Errorf("Fields = %+v, want a dateOverrides entry", fields)
		}
	})
}

func TestPageInputValidate(t *testing.T) {
	t.Run("accepts a valid page", func(t *testing.T) {
		if err := baseInput(nil).Validate(); err != nil {
			t.Errorf("Validate() = %v, want nil", err)
		}
	})

	t.Run("rejects an invalid timezone", func(t *testing.T) {
		in := baseInput(func(in *bookings.PageInput) { in.Timezone = "Not/AZone" })
		fields := fieldsOf(t, in.Validate())
		if _, ok := fields["timezone"]; !ok {
			t.Errorf("Fields = %+v, want a timezone entry", fields)
		}
	})

	t.Run("rejects out-of-range duration and buffers", func(t *testing.T) {
		cases := []func(*bookings.PageInput){
			func(in *bookings.PageInput) { in.SlotDurationMin = 10 },
			func(in *bookings.PageInput) { in.SlotDurationMin = 481 },
			func(in *bookings.PageInput) { in.BufferBeforeMin = 121 },
		}
		for _, mutate := range cases {
			in := baseInput(mutate)
			if err := in.Validate(); err == nil {
				t.Error("Validate() = nil, want a range error")
			}
		}
	})

	t.Run("rejects an invalid status", func(t *testing.T) {
		in := baseInput(func(in *bookings.PageInput) { in.Status = "archived" })
		fields := fieldsOf(t, in.Validate())
		if _, ok := fields["status"]; !ok {
			t.Errorf("Fields = %+v, want a status entry", fields)
		}
	})

	t.Run("empty status passes Validate (CreatePage ignores status; UpdatePage requires it itself)", func(t *testing.T) {
		if err := baseInput(func(in *bookings.PageInput) { in.Status = "" }).Validate(); err != nil {
			t.Errorf("Validate() = %v, want nil", err)
		}
	})

	t.Run("rejects an omitted availability (nil map) — never stored as JSON null", func(t *testing.T) {
		in := baseInput(func(in *bookings.PageInput) { in.Availability = nil })
		fields := fieldsOf(t, in.Validate())
		if fields["availability"] != "availability is required" {
			t.Errorf("Fields = %+v, want availability: availability is required", fields)
		}
	})

	t.Run("accepts an empty availability object (a page with no open days is valid, just unbookable)", func(t *testing.T) {
		in := baseInput(func(in *bookings.PageInput) { in.Availability = bookings.Availability{} })
		if err := in.Validate(); err != nil {
			t.Errorf("Validate() = %v, want nil", err)
		}
	})
}

// baseBookInput mirrors baseInput's role, for BookInput.Validate's own rule set (bookSlotSchema,
// schemas.ts): a valid visitor-facing booking request with every checked field populated.
func baseBookInput(mutate func(*bookings.BookInput)) bookings.BookInput {
	note := "See you then"
	in := bookings.BookInput{
		StartAt:  time.Now().Add(24 * time.Hour),
		Name:     "Ada Lovelace",
		Email:    "ada@example.com",
		Note:     &note,
		Timezone: "Europe/Oslo",
	}
	if mutate != nil {
		mutate(&in)
	}
	return in
}

func TestBookInputValidate(t *testing.T) {
	t.Run("accepts a valid booking", func(t *testing.T) {
		if err := baseBookInput(nil).Validate(); err != nil {
			t.Errorf("Validate() = %v, want nil", err)
		}
	})

	t.Run("name: required, trimmed, at most 80 characters", func(t *testing.T) {
		cases := []func(*bookings.BookInput){
			func(in *bookings.BookInput) { in.Name = "" },
			func(in *bookings.BookInput) { in.Name = "   " },
			func(in *bookings.BookInput) { in.Name = strings.Repeat("a", 81) },
		}
		for _, mutate := range cases {
			in := baseBookInput(mutate)
			fields := fieldsOf(t, in.Validate())
			if _, ok := fields["name"]; !ok {
				t.Errorf("Fields = %+v, want a name entry", fields)
			}
		}
		// Exactly the limit, and leading/trailing space trimmed off before counting, both pass.
		in := baseBookInput(func(in *bookings.BookInput) { in.Name = "  " + strings.Repeat("a", 80) + "  " })
		if err := in.Validate(); err != nil {
			t.Errorf("Validate() = %v, want nil", err)
		}
	})

	t.Run("email: required and must be a valid address, at most 254 characters", func(t *testing.T) {
		cases := []string{"", "not-an-email", "bob@", "Bob <bob@example.com>", strings.Repeat("a", 250) + "@a.co"}
		for _, email := range cases {
			in := baseBookInput(func(in *bookings.BookInput) { in.Email = email })
			fields := fieldsOf(t, in.Validate())
			if _, ok := fields["email"]; !ok {
				t.Errorf("email %q: Fields = %+v, want an email entry", email, fields)
			}
		}
		valid := baseBookInput(func(in *bookings.BookInput) { in.Email = "ada@example.com" })
		if err := valid.Validate(); err != nil {
			t.Errorf("Validate() = %v, want nil", err)
		}
	})

	t.Run("note: optional, trimmed, at most 1000 characters", func(t *testing.T) {
		in := baseBookInput(func(in *bookings.BookInput) { in.Note = nil })
		if err := in.Validate(); err != nil {
			t.Errorf("nil note: Validate() = %v, want nil", err)
		}

		tooLong := strings.Repeat("a", 1001)
		in = baseBookInput(func(in *bookings.BookInput) { in.Note = &tooLong })
		fields := fieldsOf(t, in.Validate())
		if _, ok := fields["note"]; !ok {
			t.Errorf("Fields = %+v, want a note entry", fields)
		}
	})

	t.Run("timezone: 1-64 characters and a real IANA zone", func(t *testing.T) {
		cases := []string{"", "Not/AZone", strings.Repeat("a", 65)}
		for _, tz := range cases {
			in := baseBookInput(func(in *bookings.BookInput) { in.Timezone = tz })
			fields := fieldsOf(t, in.Validate())
			if _, ok := fields["timezone"]; !ok {
				t.Errorf("timezone %q: Fields = %+v, want a timezone entry", tz, fields)
			}
		}
	})
}
